### PR TITLE
[CC-24379] Add folder resource and folder CRUD functionality

### PR DESCRIPTION
**Pull request by jenngeorge**
This change introduces a new `Folder` resource and adds the
relevant CRUD functionality. Support for cluster parent_id
is also added.

**Commit checklist**
- [x] Changelog
- [x] Doc gen (`make generate`)
- [x] Integration test(s)
- [x] Acceptance test(s)
- [x] Example(s)

TODO
- [x] new folder resource
- [x] support cluster parent_id
- [x] workflow example with folders, a cluster with parent_id, set folder roles
- [x] replace local cc go sdk use with the updated version (when available)

- - -
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/150)

### CODE REVIEW

**Comment by jenngeorge** [comment]
Adding to the changes in https://github.com/cockroachdb/terraform-provider-cockroach/pull/124 while @cameronnunez  is OOO.

File: internal/provider/folder_resource.go
Comment (jenngeorge) [inline internal/provider/folder_resource.go:245]: TODO: mark these api response fields required.

File: internal/provider/folder_resource_test.go
Comment (jenngeorge) [inline internal/provider/folder_resource_test.go:21]: For this test, we'll need to enable folders for the test org.

Comment (erademacher) [reply internal/provider/folder_resource_test.go:21]: Done.

**Review by jason-crl — APPROVED** [review]
LGTM!

File: internal/provider/folder_resource_test.go
Comment (jenngeorge) [reply internal/provider/folder_resource_test.go:21]: Thanks! I ran these tests using my service account in the test org, then updated the acceptance test service account's roles to include the org-level FOLDER_ADMIN role (and other FGAC admin roles). 

Screenshot 2023-08-03 at 2 37 11 PM

The serverless cluster test is also failing for the main branch, so I think it's unrelated to changes in this PR. 
Screenshot 2023-08-03 at 3 52 47 PM

**Review by erademacher — APPROVED** [review]
> Reviewed 1 of 13 files at r1, 15 of 29 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/150)* status: 16 of 32 files reviewed, 1 unresolved discussion (waiting on @cameronnunez and @jenngeorge)

*[a discussion](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/150#gh-1278211312:-Nb18jRq53_fCIRoFR6_:b-29rawy) (no related file):*
<details><summary><i>Previously, jenngeorge (Jenn Georgevich) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

TODO: mark these api response fields required.
</blockquote></details>

:lgtm: Looks great! (assuming you'll wait for the SDK update to merge)

___

**Review by jenngeorge — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/150)* status: 16 of 32 files reviewed, 1 unresolved discussion (waiting on @cameronnunez and @erademacher)

*[a discussion](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/150#gh-1278211312:-Nb1epea0Ajc92MFtf4R:b-d9dlfq) (no related file):*
<details><summary><i>Previously, erademacher (Evan Rademacher) wrote…</i></summary><blockquote>

:lgtm: Looks great! (assuming you'll wait for the SDK update to merge)
</blockquote></details>

Thanks! Yes, waiting for the SDK update before squashing commits and merging.

___

File: internal/provider/folder_resource_test.go
Comment (jenngeorge) [reply internal/provider/folder_resource_test.go:21]: Now the tests all pass, thanks for fixing that one!

**Review by jenngeorge — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/terraform-provider-cockroach/150#-:-Nbpcg6I0PfMCQbLh7bM:bz5mair)* status: 2 of 34 files reviewed, 1 unresolved discussion (waiting on @cameronnunez and @erademacher)

**Pull request merged**
